File: Huawei/简单的自动曝光.java
import java.util.Scanner;
 
/**
 * 简单的自动曝光
 * @since 2023.09.09
 * @version 0.1
 * @author Frank
 *
 */
public class ImgPixelChange {
	public static void main(String[] args) {
 
		Scanner sc = new Scanner(System.in);
		while (sc.hasNext()) {
			String input = sc.nextLine();
			String[] strNumbers = input.split( " " );
			int[] numbers = convertStrNumbers2Int( strNumbers);
			processImgPixelChange( numbers );
		}
	}
	
	private static int[] convertStrNumbers2Int( String[] numbers)
	{
		int[] ret = new int[numbers.length];
		for( int i = 0; i < numbers.length; i ++ )
		{
			ret[i] = Integer.parseInt( numbers[i] );
		}
		return ret;
	}
	
	private static void processImgPixelChange( int numbers[] )
	{
		int minDiff = Integer.MAX_VALUE;
		int AVERAGE_VALUE = 128 * numbers.length;
		int finalK = -127;
		for( int k = -127; k <= 128; k ++ )
		{
			int sum = 0;
			for( int i = 0; i < numbers.length; i ++ )
			{
				int tmpNumber = numbers[i] + k;
				if( tmpNumber < 0 )
				{
					tmpNumber = 0;
				}else if( tmpNumber > 255 )
				{
					tmpNumber = 255;
				}
				sum += tmpNumber;
			}
			int diff = Math.abs( sum - AVERAGE_VALUE );
			if( diff < minDiff )
			{
				minDiff = diff;
				finalK = k;
			}
		}		
		
		System.out.println( finalK );
	}
}

File: Huawei/简单的自动曝光.py


arr =list(map(int, input().split()))
print(arr)


import sys
def result(arr):
    minDiff = sys.maxsize
    ans = None
    # 要注意，0拉倒128的话是要加上128的，255拉倒128是要加上-127的所以才有了range(-127, 129)
    for k in range(-127, 129):
        sum = 0
        for j in range(len(arr)):
            # 先算arr[j] + k,如果小于0，那么取0，所以max(0, arr[j]+k), 
            # 如果大于255，就要取255。所以是min(max(0, arr[j]+k), 255)
            newVal = min(max(0, arr[j]+k), 255)
            sum += newVal
        diff = abs(sum/len(arr) - 128)
        # 开始与minDiff比对。
        if diff <minDiff:
            minDiff = diff
            ans = k
        elif diff == minDiff and ans is not None:
            # 如有多个整数 k 都满足，输出小的那个 k
            ans = min(ans, k)
    return ans
print(result(arr))